wal/src/message_queue_impl/region_meta.rs

```diff
@@ -0,0 +1,587 @@
+            .await;
+
+        Ok(())
+    }
+
+    /// Mark the deleted sequence number to latest next sequence number.
+    ///
+    /// NOTICE: The passed sequence number is the largest of writing
+    /// successfully, we should set the deleted one to its next one.
+    pub async fn mark_table_deleted(
+        &self,
+        table_id: TableId,
+        sequence_num: SequenceNumber,
+    ) -> Result<()> {
+        let inner = self.inner.read().await;
+        let table_meta = inner
+            .table_metas
+            .get(&table_id)
+            .with_context(|| MarkDeleted {
+                table_id,
+                msg: format!("table:{}'s meta not found while mark its deleted", table_id),
+            })?;
+        table_meta.mark_deleted(sequence_num + 1).await;
+
+        Ok(())
+    }
+
+    /// Scan the table meta entry in it and get the snapshot about tables' next
+    /// sequences.
+    ///
+    /// NOTICE: Need to freeze the whole region meta on high-level before
+    /// calling.
+    pub async fn get_snapshot(&self) -> RegionMetaSnapshot {
+        let inner = self.inner.read().await;
+        // Calc the min offset in message queue.
+        let mut entries = Vec::with_capacity(inner.table_metas.len());
+        for (table_id, table_meta) in &inner.table_metas {
+            let meta_data = table_meta.get_meta_data().await;
+            entries.push(meta_data);
+        }
+
+        RegionMetaSnapshot { entries }
+    }
+}
+
+/// Region meta data.
+#[derive(Default, Debug)]
+struct RegionMetaInner {
+    table_metas: HashMap<TableId, TableMeta>,
+}
+
+/// Wrapper for the [TableMetaInner].
+// TODO: will be made use of later.
+#[allow(unused)]
+#[derive(Debug)]
+struct TableMeta {
+    table_id: TableId,
+    /// The race condition may occur between writer thread
+    /// and background flush tread.
+    inner: Mutex<TableMetaInner>,
+}
+
+// TODO: will be made use of later.
+#[allow(unused)]
+impl TableMeta {
+    fn new(table_id: TableId) -> Self {
+        Self {
+            table_id,
+            inner: Mutex::new(TableMetaInner::default()),
+        }
+    }
+
+    async fn prepare_for_write(&self) -> SequenceNumber {
+        let inner = self.inner.lock().await;
+        inner.next_sequence_num
+    }
+
+    async fn update_after_write(&self, write_offset_range: OffsetRange, updated_num: u64) {
+        let mut inner = self.inner.lock().await;
+        let old_next_sequence_num = inner.next_sequence_num;
+        inner.next_sequence_num += updated_num;
+
+        // Update the mapping and high water mark.
+        let _ = inner
+            .start_sequence_offset_mapping
+            .insert(old_next_sequence_num, write_offset_range.start);
+        inner.current_high_watermark = write_offset_range.end + 1;
+    }
+
+    async fn mark_deleted(&self, latest_marked_deleted: SequenceNumber) {
+        let mut inner = self.inner.lock().await;
+        assert!(
```

Comment:
For this case, maybe we should return an error rather than panic.

Reply:
Maybe following error, we should return to tell : 
+ returned from the third party or os
+ returned from other module
+ caused by Distributed Inconsistency

And the pure logic error, assert maybe proper?